File: Top-Interview-Questions/Math-Easy-Collection/13-roman-to-integer.py
class Solution:
    def romanToInt(self, s: str) -> int:
        symbol = {
            "I": 1,
            "V": 5,
            "X": 10,
            "L": 50,
            "C": 100,
            "D": 500,
            "M": 1000
        }
        c = 0 # roman char index in string
        total_sum = 0 # return value
        
        """Rule
        1. if the following char is greater than previous, subtract
        2. repeat 2 every 2 chars 
        3. add up all other chars 
        """
        while c < len(s):  
            if c+1 < len(s) and symbol[s[c]] < symbol[s[c+1]]: # subtractive case, repeat every 2 chars
                total_sum = total_sum + symbol[s[c+1]] - symbol[s[c]]
                c += 2
            else:
                total_sum = total_sum + symbol[s[c]] # add up the first char and other chars except tor 
                                                 # subtractive case
                c += 1
                
        return total_sum
        
        # TC: O(1), the input is limited from 1 to 3999, but generally O(n), n is the length of the string
        # SC: O(1), the map size is only 7 cases.
        # Runtime: 40 ms, faster than 92.29%
        # Memory Usage: 13.9 MB, less than 26.91%

File: Top-Interview-Questions/Math-Easy-Collection/13-roman-to-integer.cpp
// Solution #1 Brute Force using Switch
// TC: O(1) 
// SC: O(1)
// Runtime: 12 ms, faster than 75.35%
// Memory Usage: 6.1 MB, less than 65.91%
class Solution {
    int index;
    int n;
    int res;    
    
public:
    
    int romanToInt(string s) {
        
        index = 0, res = 0;
        n = s.length()-1;
        
        while(index <= n){
            calculate(s);
            index++;
        }
        
        return res;
    }
    
    void calculate(string s){
        
        switch(s[index]){
            case 'I': checkNextofI(s); break;
            case 'V': res += 5; break;
            case 'X': checkNextofX(s); break;
            case 'L': res += 50; break;
            case 'C': checkNextofC(s); break; 
            case 'D': res += 500; break;
            case 'M': res += 1000; break;
            default: break;
        }
    }
    
    void checkNextofI(string s){
        char c = s[index+1];
        
        if(c == 'V'){
            res += 4;
            index++;
            
        }else if(c == 'X'){
            res += 9;
            index++;
            
        }else{
            res += 1;
        }
    }
    
    void checkNextofX(string s){
        char c = s[index+1];
        
        if(c == 'L'){
            res += 40;
            index++;
            
        }else if(c == 'C'){
            res += 90;
            index++;
            
        }else{
            res += 10;
        }
    }
    
    void checkNextofC(string s){
        char c = s[index+1];
        
        if(c == 'D'){
            res += 400;
            index++;
            
        }else if(c == 'M'){
            res += 900;
            index++;
            
        }else{
            res += 100;
        }        
    }
};

// Solution #2 Map
// TC: O(1) 
// SC: O(1) 7 Symbols to be stored in map
// Runtime: 12 ms, faster than 75.35%
// Memory Usage: 6.1 MB, less than 65.91%
class Solution {
    int index;
    int n;
    int res;    
    
    map<char, int> m{{'I',1}, {'V',5}, {'X', 10}, {'L',50}, {'C',100}, {'D',500}, {'M',1000}};
    
public:
    
    int romanToInt(string s) {
        
        index = 0, res = 0;
        n = s.length()-1;
        
        while(index <= n){
            
            if(index+1 <= n && m[s[index+1]] > m[s[index]]){
                res += m[s[index+1]] - m[s[index]];
                index++;
            }else{
                res += m[s[index]];
            }
            
            index++;
        }
        
        return res;
    }
    
    
};